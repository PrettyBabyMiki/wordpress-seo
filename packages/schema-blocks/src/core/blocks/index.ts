import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";
import { ValidatingBlockInstruction } from "./ValidatingBlockInstruction";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
	ValidatingBlockInstruction,
};
