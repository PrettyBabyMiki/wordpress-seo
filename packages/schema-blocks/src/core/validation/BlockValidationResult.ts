import { BlockValidation } from ".";

/**
 * Contains the result of a block validation.
 */
export class BlockValidationResult {
	/**
	 * The name of the validated block.
	 */
	public name: string;

	/**
	 * The clientId of the validated block.
	 */
	public clientId: string;

	/**
	 * The validation result;
	 */
	public result: BlockValidation;

	/**
	 * The validation issues for this block's innerblocks or attributes, if any.
	 */
	public issues: BlockValidationResult[]

	/**
	 * @param clientId The clientId of the validated block.
	 * @param name     The name of the validated block.
	 * @param result   The validation result.
	 */
	constructor( clientId: string, name: string, result: BlockValidation ) {
		this.name = name;
		this.clientId = clientId;
		this.result = result;
		this.issues = [];
	}
}
