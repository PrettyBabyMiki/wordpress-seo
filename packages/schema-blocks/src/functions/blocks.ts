import { dispatch, select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}

/**
 * Removes a block from the editor.
 *
 * @param {string} clientId The client id of the block to remove.
 */
export function removeBlock( clientId: string ) {
	dispatch( "core/block-editor" ).removeBlock( clientId );
}
