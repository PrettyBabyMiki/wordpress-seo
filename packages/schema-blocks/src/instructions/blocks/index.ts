import "./Block";
import "./ClassName";
import "./InnerBlocks";
import "./RichText";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./VariableTagRichText";
