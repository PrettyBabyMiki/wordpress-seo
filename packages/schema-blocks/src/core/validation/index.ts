import { BlockValidation } from "./BlockValidation";
import { BlockValidationResult } from "./BlockValidationResult";
import { RequiredBlockOption } from "./RequiredBlockOption";
import { RequiredBlock } from "./RequiredBlock";
import { RecommendedBlock } from "./RecommendedBlock";
import { SuggestedBlockProperties } from "./SuggestedBlockProperties";

export { BlockValidation, BlockValidationResult, RequiredBlockOption, RequiredBlock, RecommendedBlock, SuggestedBlockProperties };
