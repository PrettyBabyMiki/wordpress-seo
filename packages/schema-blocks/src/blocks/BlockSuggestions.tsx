import { ReactElement } from "react";
import { BlockInstance, createBlock } from "@wordpress/blocks";
import { createElement } from "@wordpress/element";
import { getInnerblocksByName, insertBlock } from "../functions/innerBlocksHelper";
import { getBlockType } from "../functions/BlockHelper";
import { PanelBody } from "@wordpress/components";
import { SuggestedBlockProperties } from "../core/validation/SuggestedBlockProperties";

type BlockSuggestionAddedDto = {
	blockTitle: string;
}

type BlockSuggestionDto = {
	blockTitle: string;
	blockName: string;
	blockClientId: string;
}

/**
 * Renders a block suggestion with the possibility to add one.
 *
 * @param {string}   blockTitle    The title to show.
 * @param {string}   blockName     The name of the block to add.
 * @param {string}   blockClientId The clientId of the target to add the block to.
 *
 * @returns {ReactElement} The rendered block suggestion.
 */
function BlockSuggestion( { blockTitle, blockName, blockClientId }: BlockSuggestionDto ): ReactElement {
	/**
	 * Onclick handler for the remove block.
	 */
	const addBlockClick = () => {
		const blockToAdd = createBlock( blockName );
		insertBlock( blockToAdd, blockClientId );
	};

	return (
		<li className="yoast-block-suggestion">
			{ blockTitle }
			<button className="yoast-block-suggestion-button" onClick={ addBlockClick }> Add </button>
		</li>
	);
}

/**
 * Renders a block suggestion that is already added
 *
 * @param {string} blockTitle The block title.
 *
 * @returns {ReactElement} The rendered element.
 */
function BlockSuggestionAdded( { blockTitle }: BlockSuggestionAddedDto ): ReactElement {
	return (
		<li className="yoast-block-suggestion yoast-block-suggestion--added">
			{ blockTitle }
			<span className="yoast-block-suggestion-checkmark"> OK </span>
		</li>
	);
}

/**
 * Renders a list with the required/recommended block names and an button to add/remove one.
 *
 * @param {string} sidebarTitle                The title of the sidebar section.
 * @param {BlockInstance} block                The block to render the list for.
 * @param {SuggestedBlockProperties[]} blocks The required/recommended blocks.
 *
 * @returns {ReactElement} The rendered block.
 */
export default function RequiredBlocks( sidebarTitle: string, block: BlockInstance, blocks: SuggestedBlockProperties[] ): ReactElement {
	// Retrieve a list with names.
	const relevantBlockNames = blocks
		.filter( relevantBlock => typeof getBlockType( relevantBlock.name ) !== "undefined" )
		.map( relevantBlock => relevantBlock.name );

	// When there are no required blocknames, just return.
	if ( relevantBlockNames.length === 0 ) {
		return null;
	}

	const findPresentBlocks = getInnerblocksByName( block, relevantBlockNames );
	const presentBlockNames = findPresentBlocks.map( presentBlock => presentBlock.name );

	return (
		<PanelBody key={ block.clientId }>
			<div className="yoast-block-sidebar-title">{ sidebarTitle }</div>
			<ul className="yoast-block-suggestions">
				{
					relevantBlockNames.map( ( blockName: string, index: number ) => {
						const blockType = getBlockType( blockName );

						if ( presentBlockNames.includes( blockName ) ) {
							return <BlockSuggestionAdded key={ index } blockTitle={ blockType.title } />;
						}

						return <BlockSuggestion
							key={ index }
							blockTitle={ blockType.title }
							blockName={ blockName }
							blockClientId={ block.clientId }
						/>;
					} )
				}
			</ul>
		</PanelBody>
	);
}
