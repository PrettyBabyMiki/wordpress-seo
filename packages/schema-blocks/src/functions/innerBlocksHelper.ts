import { BlockInstance } from "@wordpress/blocks";
import { select } from "@wordpress/data";
import recurseOverBlocks from "./blocks/recurseOverBlocks";

/**
 * Searches recursively in the inner blocks to get all instances of blocks whose name occurs in blockNames.
 *
 * @param blockNames     The names of the blocks you're searching for.
 * @param blockInstances The array of blocks you're searching in.
 *
 * @returns {BlockInstance[]} The block instances that have a name that occurs in blockNames.
 */
function getInnerblocksByName( blockNames: string[], blockInstances: BlockInstance[] ): BlockInstance[] {
	const foundBlocks: BlockInstance[] = [];

	recurseOverBlocks( blockInstances, ( block: BlockInstance ) => {
		// Checks if the current block is one of the required types.
		if ( blockNames.includes( block.name ) ) {
			foundBlocks.push( block );
		}
	} );

	return foundBlocks;
}

/**
 * Gets the inner blocks of the block with the given clientId from the core/block-editor store.
 *
 * @param clientId The clientId of the block whose InnerBlocks you want.
 *
 * @returns {BlockInstance[]} The block's inner blocks.
 */
function getInnerBlocks( clientId: string ): BlockInstance[] {
	/* istanbul ignore next */
	return select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
}

export { getInnerBlocks, getInnerblocksByName };
