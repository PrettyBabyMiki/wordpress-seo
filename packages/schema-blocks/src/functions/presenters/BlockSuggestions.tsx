import { BlockInstance, createBlock } from "@wordpress/blocks";
import { getBlockType } from "../BlockHelper";
import { getInnerblocksByName, insertBlock } from "../innerBlocksHelper";
import { PanelBody } from "@wordpress/components";
import { createElement } from "@wordpress/element";
import { ReactElement } from "react";

type BlockSuggestionAddedDto = {
	blockTitle: string;
}

type BlockSuggestionDto = {
	blockTitle: string;
	blockName: string;
	blockClientId: string;
}

interface BlockSuggestionsProps {
	title: string;
	block: BlockInstance;
	suggestions: string[];
}

/**
 * Renders a block suggestion with the possibility to add one.
 *
 * @param {string}   blockTitle    The title to show.
 * @param {string}   blockName     The name of the block to add.
 * @param {string}   blockClientId The clientId of the target to add the block to.
 *
 * @returns {ReactElement} The rendered block suggestion.
 */
function BlockSuggestion( { blockTitle, blockName, blockClientId }: BlockSuggestionDto ): ReactElement {
	/**
	 * Onclick handler for the remove block.
	 */
	const addBlockClick = () => {
		const blockToAdd = createBlock( blockName );
		insertBlock( blockToAdd, blockClientId );
	};

	return (
		<li className="yoast-block-suggestion">
			{ blockTitle }
			<button className="yoast-block-suggestion-button" onClick={ addBlockClick }> Add </button>
		</li>
	);
}

/**
 * Renders a block suggestion that is already added
 *
 * @param {string} blockTitle The block title.
 *
 * @returns {ReactElement} The rendered element.
 */
function BlockSuggestionAdded( { blockTitle }: BlockSuggestionAddedDto ): ReactElement {
	return (
		<li className="yoast-block-suggestion yoast-block-suggestion--added">
			{ blockTitle }
			<span className="yoast-block-suggestion-checkmark"> OK </span>
		</li>
	);
}

export default function BlockSuggestions( { title, block, suggestions }: BlockSuggestionsProps ) {
	const suggestedBlockNames = suggestions
		.filter( suggestedBlock => typeof getBlockType( suggestedBlock ) !== "undefined" )
		.map( suggestedBlock => suggestedBlock );

	// When there are no suggestions, just return.
	if ( suggestedBlockNames.length === 0 ) {
		return null;
	}

	const findPresentBlocks = getInnerblocksByName( block, suggestedBlockNames );
	const presentBlockNames = findPresentBlocks.map( presentBlock => presentBlock.name );

	return (
		<PanelBody key={ block.clientId }>
			<div className="yoast-block-sidebar-title">{ title }</div>
			<ul className="yoast-block-suggestions">
				{
					suggestedBlockNames.map( ( blockName: string, index: number ) => {
						const blockType = getBlockType( blockName );

						if ( presentBlockNames.includes( blockName ) ) {
							return <BlockSuggestionAdded key={ index } blockTitle={ blockType.title } />;
						}

						return <BlockSuggestion
							key={ index }
							blockTitle={ blockType.title }
							blockName={ blockName }
							blockClientId={ block.clientId }
						/>;
					} )
				}
			</ul>
		</PanelBody>
	);
}
