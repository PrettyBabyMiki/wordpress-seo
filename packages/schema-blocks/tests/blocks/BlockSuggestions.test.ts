import { BlockInstance, createBlock } from "@wordpress/blocks";
import * as renderer from "react-test-renderer";
import { mount } from "enzyme";
import { BlockValidationResult, RequiredBlock } from "../../src/core/validation";
import BlockSuggestions from "../../src/blocks/BlockSuggestions";
import { insertBlock } from "../../src/functions/innerBlocksHelper";
import { BlockPresence } from "../../src/core/validation/BlockValidationResult";

jest.mock( "@wordpress/blocks", () => {
	return {
		createBlock: jest.fn(),
	};
} );

jest.mock( "../../src/functions/validators", () => {
	return {
		getValidationResult: jest.fn( () => {
			return new BlockValidationResult( "1", "yoast/valid-block", -1, BlockPresence.Required, "Is not that present" );
		} ),
		getAllDescendantIssues: jest.fn( () => {
			return [
				new BlockValidationResult( "123", "yoast/added-to-content-valid", 1, BlockPresence.Required, "Is present" ),
			];
		} ),
	};
} );

jest.mock( "../../src/functions/BlockHelper", () => {
	return {
		getBlockType: jest.fn( ( blockName: string )  => {
			if ( blockName === "yoast/nonexisting" ) {
				// eslint-disable-next-line no-undefined
				return undefined;
			}

			return {
				title: "The required block",
			};
		} ),
	};
} );

jest.mock( "../../src/functions/innerBlocksHelper", () => {
	return {
		insertBlock: jest.fn(),
		getInnerblocksByName: jest.fn( ()  => {
			return [
				{
					name: "yoast/added-to-content",
				},
			];
		} ),
	};
} );

describe( "The required blocks in the sidebar", () => {
	it( "doesn't have the required block being registered as a block", () => {
		const block = { innerBlocks: [] } as BlockInstance;
		const requiredBlocks = [
			{
				name: "yoast/nonexisting",
				option: "One",
			} as RequiredBlock,
		];

		const actual = BlockSuggestions( "Required blocks", block, requiredBlocks );

		expect( actual ).toBe( null );
	} );

	it( "renders the required block as an added one", () => {
		const block = { clientId: "1", innerBlocks: [] } as BlockInstance;
		const requiredBlocks = [
			{
				name: "yoast/added-to-content",
				option: "One",
			} as RequiredBlock,
			{
				name: "yoast/added-to-content-valid",
				option: "One",
			} as RequiredBlock,
		];

		const tree = renderer
			.create( BlockSuggestions( "Required blocks", block, requiredBlocks ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the required block as a non-added one", () => {
		const block = { innerBlocks: [] } as BlockInstance;
		const requiredBlocks = [
			{
				name: "yoast/non-added-to-content",
				option: "One",
			} as RequiredBlock,
		];

		const tree = renderer
			.create( BlockSuggestions( "Required blocks", block, requiredBlocks ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "should call the function to add the block when the button is clicked.", () => {
		const block = { innerBlocks: [], clientId: "1" } as BlockInstance;
		const requiredBlocks = [
			{
				name: "yoast/non-added-to-content",
				option: "One",
			} as RequiredBlock,
		];

		const tree = mount( BlockSuggestions( "Required blocks", block, requiredBlocks ) );

		const addButton = tree.find( "button" ).first();

		addButton.simulate( "click" );

		expect( createBlock ).toHaveBeenCalled();
		expect( insertBlock ).toHaveBeenCalled();
	} );
} );
