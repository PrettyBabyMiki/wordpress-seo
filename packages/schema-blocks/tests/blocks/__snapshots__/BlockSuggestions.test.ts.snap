// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The required blocks in the sidebar renders the required block as a non-added one 1`] = `
<div
  className="components-panel__body is-opened"
>
  <div
    className="yoast-block-sidebar-title"
  >
    Required blocks
  </div>
  <ul
    className="yoast-block-suggestions"
  >
    <li
      className="yoast-block-suggestion"
    >
      The required block
      <button
        className="yoast-block-suggestion-button"
        onClick={[Function]}
      >
         Add 
      </button>
    </li>
  </ul>
</div>
`;

exports[`The required blocks in the sidebar renders the required block as an added one 1`] = `
<div
  className="components-panel__body is-opened"
>
  <div
    className="yoast-block-sidebar-title"
  >
    Required blocks
  </div>
  <ul
    className="yoast-block-suggestions"
  >
    <li
      className="yoast-block-suggestion yoast-block-suggestion--added"
    >
      The required block
      <span
        className="yoast-block-suggestion-checkmark"
      >
         OK 
      </span>
    </li>
  </ul>
</div>
`;
